### PR TITLE
Handle 2xx as success for OTLP HTTP trace and metric exporters

### LINKED ISSUE #4363: Exporting fails if a 204 No Content is returned
### Description

The current http exporters [expect a http.StatusOK (200) response code](https://github.com/open-telemetry/opentelemetry-go/blob/d423bd4cf28f65bdcbd1b1d83d0f59c4388f37b0/exporters/otlp/otlpmetric/otlpmetrichttp/client.go#L180-L204). But this is not required by the spec and also not [something the collector requires](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/otlphttpexporter/otlp.go#L135-L137).

Trying to use the exporter in a test fails with:

```
failed to upload metrics: failed to send metrics to http://localhost:9090/api/v1/otlp/v1/metrics: 204 No Content
```

### PR DESCRIPTION
**Pull request by dmathieu**
Closes #4363.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -17,6 +17,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - OTLP Metrics Exporter now supports the `OTEL_EXPORTER_OTLP_METRICS_TEMPORALITY_PREFERENCE` environment variable. (#4287)
 - Add `WithoutCounterSuffixes` option in `go.opentelemetry.io/otel/exporters/prometheus` to disable addition of `_total` suffixes. (#4306)
 - Add info and debug logging to the metric SDK. (#4315)
+- Handle 204 HTTP status responses in `go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp` and `go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp`. (#4365)
```
Comment (pellared) [inline CHANGELOG.md:20]: Isn't the specification requiring to return `200 OK`?

https://github.com/open-telemetry/opentelemetry-proto/blob/main/docs/specification.md#otlphttp-response

If so then, we should first update the specification.

Comment (dmathieu) [reply CHANGELOG.md:20]: It also mentions that any other HTTP status code should be handled per the HTTP spec.
See my comment in https://github.com/open-telemetry/opentelemetry-go/issues/4363#issuecomment-1651464538

Comment (pellared) [reply CHANGELOG.md:20]: I see. The spec mentions it here: https://github.com/open-telemetry/opentelemetry-proto/blob/main/docs/specification.md#all-other-responses

Comment (pellared) [reply CHANGELOG.md:20]: The problem is that the specification requires

> the server MUST respond with HTTP 200 OK

While 

> All other HTTP responses that are not explicitly listed in this document should be treated according to HTTP specifications.

is not really specific and it does not seem to be normative.

Have you checked how OTLP exporters in other SDKs (e.g. in Java, .NET, Python) are handling it?

Comment (dmathieu) [reply CHANGELOG.md:20]: Python accepts 200 and 202: https://github.com/open-telemetry/opentelemetry-python/blob/main/exporter/opentelemetry-exporter-otlp-proto-http/src/opentelemetry/exporter/otlp/proto/http/metric_exporter/__init__.py#L191

Java accepts 200 to 299: https://github.com/open-telemetry/opentelemetry-java/blob/main/exporters/common/src/main/java/io/opentelemetry/exporter/internal/http/HttpExporter.java#L85

DotNet relies on `EnsureSuccessStatusCode`, which seems to accept 200 to 299: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/BaseOtlpHttpExportClient.cs#L58C27-L58C50

Node accepts 200 to 299: https://github.com/open-telemetry/opentelemetry-js/blob/main/experimental/packages/otlp-exporter-base/src/platform/node/util.ts#L102

Comment (pellared) [reply CHANGELOG.md:20]: Collector accepts 200 to 299: https://github.com/open-telemetry/opentelemetry-collector/blob/56135230d1d998608cc7c61250b28b64ffdd620b/exporter/otlphttpexporter/otlp.go#L135-L137

Comment (dmathieu) [reply CHANGELOG.md:20]: That's what I was starting to think too.

File: exporters/otlp/otlpmetric/internal/otest/client.go [hunk 2]
```diff
@@ -272,6 +272,33 @@ func RunClientTests(f ClientFactory) func(*testing.T) {
...
+					defer func(orig otel.ErrorHandler) {
+						otel.SetErrorHandler(orig)
+					}(otel.GetErrorHandler())
+
+					errs := []error{}
+					eh := otel.ErrorHandlerFunc(func(e error) { errs = append(errs, e) })
+					otel.SetErrorHandler(eh)
+
+					require.NoError(t, client.UploadMetrics(ctx, nil))
+					require.NoError(t, client.Shutdown(ctx))
```
Comment (pellared) [inline exporters/otlp/otlpmetric/internal/otest/client.go:296]: I think this should be added invoked as part of `t.Cleanup` (or deferred).

File: exporters/otlp/otlpmetric/internal/otest/client.go [hunk 3]
```diff
@@ -272,6 +272,33 @@ func RunClientTests(f ClientFactory) func(*testing.T) {
 			want := fmt.Sprintf("%s (%d metric data points rejected)", msg, n)
 			assert.ErrorContains(t, errs[0], want)
 		})
+
+		t.Run("Other HTTP success", func(t *testing.T) {
+			for i := 201; i <= 299; i++ {
```
Comment (pellared) [inline exporters/otlp/otlpmetric/internal/otest/client.go:277]: nit: Can we rename `i` to `code`?

File: exporters/otlp/otlpmetric/internal/otest/client.go [hunk 4]
```diff
@@ -272,6 +272,33 @@ func RunClientTests(f ClientFactory) func(*testing.T) {
...
+					}(otel.GetErrorHandler())
+
+					errs := []error{}
+					eh := otel.ErrorHandlerFunc(func(e error) { errs = append(errs, e) })
+					otel.SetErrorHandler(eh)
+
+					require.NoError(t, client.UploadMetrics(ctx, nil))
+					require.NoError(t, client.Shutdown(ctx))
+
+					require.Equal(t, 0, len(errs))
```
Comment (pellared) [inline exporters/otlp/otlpmetric/internal/otest/client.go:298]: I think we can change `require` to `assert` as any of the assertions does not need to interrupt the test.

File: exporters/otlp/otlptrace/otlptracehttp/client_test.go [hunk 5]
```diff
@@ -401,3 +401,34 @@ func TestPartialSuccess(t *testing.T) {
 	require.Contains(t, errs[0].Error(), "partially successful")
 	require.Contains(t, errs[0].Error(), "2 spans rejected")
 }
+
+func TestOtherHTTPSuccess(t *testing.T) {
+	for i := 201; i <= 299; i++ {
```
Comment (pellared) [inline exporters/otlp/otlptrace/otlptracehttp/client_test.go:406]: nit: Can we rename `i` to `code`?

File: exporters/otlp/otlpmetric/internal/otest/client.go (hunk 2 above)
Comment (dmathieu) [reply exporters/otlp/otlpmetric/internal/otest/client.go:296]: I can do so, but then we can't check for Shutdown errors anymore. And every other test also performs Shutdown synchronously to check for errors.

Comment (pellared) [reply exporters/otlp/otlpmetric/internal/otest/client.go]: You can still assert in defer and t.Cleanup 😉

**Review by pellared — APPROVED** [review]

**Comment by pellared** [comment]
Please update the PR title and description.

**Review by MrAlias — APPROVED** [review]

**Review by dashpole — APPROVED** [review]

**Pull request merged**